### PR TITLE
T284598 gutenberg search format

### PR DESCRIPTION
**Pull request by hueitan**
Phabricator Ticket: https://phabricator.wikimedia.org/T284598

The `prefix` search feature is implemented, the previous temp button of `Add` `Remove` are removed.

- [x] Mobile
- [x] prefix search
- [x] fulltext search
- [x] RTL
- [x] Keyboard
- [x] UI final tune

| LTR | RTL | Fulltext Search |
| --- | --- | --- |
| image | image | image |

### CODE REVIEW

File: src/link/inline.js [hunk 1]
```diff
@@ -1,60 +1,154 @@
...
+				{ title && (
+					<Button
+						onClick={ () => {
+							setTitle( '' );
+						} }
+						className="wikipediapreview-edit-inline-search-close"
+					/>
+				) }
 			</div>
+			{ searchList?.length ? (
```
[the remark this answers is not in the document]
Comment (hueitan) [reply src/link/inline.js]: same to `searchList && searchList.length`, but I'm removing this now as it is not necessary here.

File: src/link/api.js [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+			} )
+			.join( '&' )
+	);
+};
+
+const request = ( url, callback ) => {
+	abortFunctions.forEach( ( x ) => x && x.abort() );
+	abortFunctions = [];
+
+	// eslint-disable-next-line no-undef
```
Comment (stephanebisson) [inline src/link/api.js:103]: I think the proper way to silent this warning is by telling eslint that this code will run in a browser so it understands that `XMLHttpRequest`, `document`, `window` etc are globals. 


In `.eslintrc`:
```
"env": {
        "browser": true
    }
```

File: src/link/inline.js [hunk 3]
```diff
@@ -1,60 +1,154 @@
...
 	const [ lang, setLang ] = useState( activeAttributes.lang );
+	const [ searchList, setSearchList ] = useState( [] );
+	const [ hoveredIndex, setHoverIndex ] = useState( -1 );
+	const onItemHovered = ( index ) => {
+		if ( ! searchList.length ) return;
+		const className = 'wikipediapreview-edit-inline-list-item';
+
+		if ( hoveredIndex >= 0 ) {
+			document
+				.querySelectorAll( `.${ className }` )
```
Comment (stephanebisson) [inline src/link/inline.js:30]: It's not really the react way to access the DOM like this.

You're already keeping the hoveredIndex in a state so the code that renders the list items can add the `hovered` class to the right item. There's a bunch or re-rendering anyway.

File: src/link/inline.js [hunk 4]
```diff
@@ -1,60 +1,154 @@
...
+						} }
+						className="wikipediapreview-edit-inline-search-close"
+					/>
+				) }
 			</div>
+			{ searchList.length ? (
+				<div className="wikipediapreview-edit-inline-list">
+					{ searchList.map( ( item, index ) => {
+						return (
+							// eslint-disable-next-line jsx-a11y/click-events-have-key-events
```
Comment (stephanebisson) [inline src/link/inline.js:97]: What is `jsx-a11y/click-events-have-key-events` trying to accomplish?

**Review by stephanebisson — CHANGES_REQUESTED** [review]
When you clear the search box with backspace it keeps the previous search results. It should probably go back to the initial (blank) state.

**Comment by hueitan** [comment]
> When you clear the search box with backspace it keeps the previous search results. It should probably go back to the initial (blank) state.

Thanks for the feedback, all should be updated now.

**Comment by stephanebisson** [comment]
> > When you clear the search box with backspace it keeps the previous search results. It should probably go back to the initial (blank) state.
> 
> Thanks for the feedback, all should be updated now.

Doesn't work for me. The only way I can clear the result is by typing a space character.

**Review by stephanebisson — CHANGES_REQUESTED** [review]
We should be able to do the following:
1. type in "Paris"
2. type ENTER without selecting any entry from the search results
since "paris" is a valid search result, it should just work, now it errors out

**Comment by hueitan** [comment]
> We should be able to do the following:
> 
> 1. type in "Paris"
> 2. type ENTER without selecting any entry from the search results
>    since "paris" is a valid search result, it should just work, now it errors out

This is available now through the ENTER key,

the only thing missing clarification is the behaviour of how the typing/keyboard/mouse trigger the hover style/selection, should it behave the same as wikipedia search feature?

**Comment by stephanebisson** [comment]
> the only thing missing clarification is the behaviour of how the typing/keyboard/mouse trigger the hover style/selection, should it behave the same as wikipedia search feature?

I think it works well as it is. I'll merge it so it can move forward to QA and design review.

**Pull request merged**